rustls/src/server/mod.rs

```diff
@@ -324,113 +337,210 @@ impl ServerConnection {
+    ///
+    /// For more details, refer to [`CommonState::wants_read()`].
+    pub fn wants_read(&self) -> bool {
+        self.inner
             .as_ref()
-            .ok_or(Error::HandshakeNotComplete)
-            .and_then(|st| st.export_keying_material(output, label, context))
+            .map(|conn| conn.common_state.wants_read())
+            .unwrap_or(false)
     }
 
-    fn negotiated_cipher_suite(&self) -> Option<SupportedCipherSuite> {
-        self.common.get_suite()
+    /// Read TLS content from `rd`.
+    ///
+    /// Returns an error if this `Acceptor` has already yielded an [`Accepted`]. For more details,
+    /// refer to [`Connection::read_tls()`].
+    pub fn read_tls(&mut self, rd: &mut dyn io::Read) -> Result<usize, io::Error> {
+        match &mut self.inner {
+            Some(conn) => conn.read_tls(rd),
+            None => Err(io::Error::new(
+                io::ErrorKind::Other,
+                "acceptor cannot read after successful acceptance",
+            )),
+        }
     }
 
-    fn writer(&mut self) -> Writer {
-        Writer::new(self)
-    }
+    /// Check if a `ClientHello` message has been received.
+    ///
+    /// Returns an error if the `ClientHello` message is invalid or if the acceptor has already
+    /// yielded an [`Accepted`]. Returns `Ok(None)` if no complete `ClientHello` has been received
+    /// yet.
+    pub fn accept(&mut self) -> Result<Option<Accepted<'_>>, Error> {
+        let mut connection = match self.inner.take() {
+            Some(conn) => conn,
+            None => {
+                return Err(Error::General(
+                    "cannot accept after successful acceptance".into(),
+                ));
+            }
+        };
 
-    fn reader(&mut self) -> Reader {
-        self.common.reader()
+        match connection.first_handshake_message() {
+            Ok(msg) => self.message = msg,
+            Err(e) => {
+                self.inner = Some(connection);
+                return Err(e);
+            }
+        }
+
+        let message = match self.message.as_ref() {
+            Some(msg) => msg,
+            None => {
+                self.inner = Some(connection);
+                return Ok(None);
+            }
+        };
+
+        let (client_hello, sig_schemes) = hs::process_client_hello(
+            message,
+            false,
+            &mut connection.common_state,
+            &mut connection.data,
+        )?;
+
+        Ok(Some(Accepted {
+            connection,
+            message,
+            client_hello,
+            sig_schemes,
+        }))
     }
 }
 
-impl PlaintextSink for ServerConnection {
-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        Ok(self.send_some_plaintext(buf))
-    }
+/// Represents a `ClientHello` message received through the [`Acceptor`].
+///
+/// Contains the state required to resume the connection through [`Accepted::into_connection()`].
+pub struct Accepted<'a> {
```

Comment:
That lifetime might make things a bit tricky for applications which really do an async lookup and have to pause the handshake for it. But that's just a feeling - I recommend testing it out in a more complete example example, e.g. one which uses `tokio::spawn_blocking` for reading certs.

Reply:
I'm hoping @jeromegn will do that.

We could do without it, but that would involve reparsing the `ClientHello` message -- I thought it might be nice to avoid the overhead.

Reply:
Any kind of validation would be nice. If `Accepted` could not be hold due to whatever lifetime issues, the workaround with this kind of API is probably to drop `Accepted` and get it back later. But that might not work due to `accept()` taking state out of the connnection, and so dropping `Accepted` would not allow to resume the handshake?

Reply:
Yeah, that wouldn't work. But as I said, we could just hold the `Message` in the `Accepted` while dropping the `client_hello` (instead cloning the ALPN protocols we need to build the `ClientHello`), then we'd reparse the `ClientHello` in `Accepted::into_connection()`.

Reply:
It could also hold the message and parsed clienthello as part of `Acceptor`, and either remove `Acceptor`  or make it just a tiny shim which retries that info? Then creating `Accepted` could be idempotent. As long as the ClientHello is there, you will get it again.

But that's just a solution before understanding whether its really a problem - integrating this plus async cert lookup in something like tokio-rustls should tell us.

Reply:
Ah yeah, I guess we wouldn't actually have to reparse the message, just extract the `ClientHelloPayload` again. Still, we need both the `Message` and the `ClientHelloPayload`, so that would just be more cloning. I think the borrowing solution is preferable as long as the lifetime doesn't get in the way too much.

Reply:
> I'm hoping @jeromegn will do that.

Thanks for including me here, but I'm barely following! I was hoping I could convert `tokio-rustls` to use this branch and get rid of our double-parsing of Client Hello for our async needs.

I can only spend time on this in short bursts. Somebody else is probably better suited than me to do the heavy-lifting if you want this to go forward faster 😄

Reply:
I did indeed end up getting right of the lifetime here, since it caused problems downstream.